state_machine/applications/flight/lib/priority_queue.py

```diff
@@ -0,0 +1,75 @@
+# Based on CPython's heapq
+# Converted to a max heap, and greatly simplified
+# https://github.com/python/cpython/blob/3.10/Lib/heapq.py
+""" Priority Queue Algorithm
```

Comment:
Why do we need to use a heap? It seems to me that a linked list queue, or a fixed-size array would be totally sufficient.

Reply:
So that we prioritize high priority transmissions such as beacons over other lower priority transmissions. It allows us to efficiently transmit what we deem to be the highest priority message. I believe this is useful because we have many different levels of priority for different message such as beacons, images, errors, state changes, logs, etc... And this creates a simple message ordering system

Reply:
I guess I'm still not sure why it's necessary to use a tree since we are only ever going to be transmitting the highest priority thing. So, just sticking the packet in the appropriate place in a list would accomplish the same thing.

Reply:
Trees are good for randomly accessing things, but this requires an O(n) sort step every time something is added. If I'm remembering my data structures correctly, a linked list would be O(1).

Reply:
I absolutely agree with you, trees would not be the correct solution. We actually are not using trees at all. We are using heaps which are just lists with a pretty loose data structure invariant. 

What we need is a data structure that allows us to quickly access the highest priority element, and fast insertion/deletion from said data structure. 

One option is a sorted binary tree which has:
- O(logn) insertion, deletion, and highest priority element access
- Memory overhead from all the pointers needed to have a tree
A sorted list
- O(1) amortized accessing highest priority element and deletion
- O(n) insertion
- Little memory overhead
Linked list (maybe this is what you meant)
- O(1) accessing highest priority element and deletion
- O(n) insertion
- Memory overhead from all the pointers
All this is why I chose a heap, which is a list with a much looser invariant than sorting which allows for:
- O(logn) amortized insertion/deletion
- O(1) highest priority element access
- is still a list so has virtually no memory overhead